### PR TITLE
feat(workspace-tools): support `--jobs 1` and `--jobs unlimited`

### LINKED ISSUE #3433: [Feature] `workspaces foreach` `--jobs` should accept 1 and have method to specify "infinity"
- [x] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

This is two closely-related feature requests in one. I can split this ticket into two if you'd like, but I figured it was less noisy this way.

#### setting jobs to 1

I use partial script definitions to make it easier to run things like `foreach` in a suitably preconfigured way:

```json
"scripts": {
  "backend:run": "yarn workspaces foreach ${FOREACH_OPTIONS:-} -pivR --from '@foo/{bar,baz}' run",
  "backend:build": "yarn backend:run build",
  "backend:codegen": "FOREACH_OPTIONS='--jobs=1' yarn backend:run codegen"
}
```

As you can see, in this example, I'd like to default to parallelizing the tasks unless they opt-out. (There are about a dozen more scripts following this basic pattern.)

Without being able to specify `--jobs=1`, I have to un-DRY my script definitions (by inlining `backend:run` into `backend:codegen` but without `-p`) or change it to be serial by default, and update all the _other_ script definitions to specify `FOREACH_OPTIONS`.

Some prior art: [Lerna permits passing `--concurrency=1`](https://github.com/lerna/lerna/blob/6cb8ab2d4af7ce25c812e8fb05cd04650105705f/core/global-options/README.md#--concurrency); [npm-run-all permits passing `--max-parallel=1`](https://github.com/mysticatea/npm-run-all/blob/master/docs/npm-run-all.md) (though the docs don't mention it, I tried it/read the code and it does).

#### setting jobs to infinity

I have repo-wide "watch"-style tasks. Since they're (1) long-running, (2) generally idle and (3) all need to be running at the same time, I do this by specifying a ridiculous number of jobs:

```
yarn workspaces foreach -piv --jobs 1000 run watch
```

It would be nice if there were some way to make it a little more obvious what I'm trying to do here. Plus, without peeking into the implementation, I don't know if I'm asking Yarn to spin up 1000 workers (which would be undesirable, to say the least).

Prior art: [npm-run-all defaults to this behavior](https://github.com/mysticatea/npm-run-all/blob/master/docs/npm-run-all.md) (ctrl-f "max-parallel", though I would not want it to be default). IIRC Lerna used to do this a long time ago, but currently it has the same behavior as Yarn (more or less).

**Describe the solution you'd like**

For setting jobs to one, simply relaxing the validation to permit it.

For setting jobs to infinity, I would _not_ want this to be default, but instead a flag. I see a few versions of this flag:

- `--max-jobs` (or a similar name) which either (1) cannot be provided with `--jobs` or (2) loses to explicit `--jobs` (perhaps with a warning)
- `--jobs=0`, which doesn't require a new flag or defining any interactions with said flag, but is a bit weird (though some old-school programs use this style, so it's not totally foreign)
- `--jobs=infinite`, which I think is the best of both worlds

Neither of these are breaking changes, as they are both strictly additive.

**Describe the drawbacks of your solution**

The only drawback I can think of for allowing settings jobs to 1 is someone erroneously configuring parallelization to 1 when otherwise Yarn would have prevented it. In general, serializing when you're you'd normally parallelize is still correct, though slow, so this should be safe. If we're really concerned about this, we could issue a warning (though I'd rather not to avoid polluting the output with something the user almost certainly did on purpose).

I don't think there are any drawbacks for allowing setting jobs to infinity, given that it's opt-in, explicit and (in 2/3 cases above) very self-evident what Yarn is being asked to do.

**Describe alternatives you've considered**

This is a request to change a core behavior and as such is not appropriate for a plugin.

Neither of these issues is blocking, but hopefully the workarounds described above demonstrate why it would be preferable to have these behaviors be properly supported.

### PR DESCRIPTION
**Pull request by seansfkelley**
**What's the problem this PR addresses?**

Permits users to specify `--jobs 1` and `--jobs unlimited` to `workspaces foreach` command.

`--jobs 1` takes precedence over `--parallel` and `--interlace`, disabling both. See the linked ticket below for justification.

`--jobs unlimited` kicks off every task simultaneously. It's intended to be used with watch tasks in combination with `--interlace`.

Resolves #3433.

**How did you fix it?**

- Change the argument validation to permit two new option values.
- Rearrange the way that concurrency/parallel/interlace arguments are initialized such that a calculated concurrency of 1 can override the parallel/interlace behavior to avoid extra work.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

File: .pnp.cjs [hunk 1]
```diff
@@ -46001,9 +46001,17 @@ async function copyFolder(prelayout, postlayout, updateTime, destinationFs, dest
   let updated = false;
 
   if (destinationStat === null) {
-    prelayout.push(async () => destinationFs.mkdirPromise(destination, {
-      mode: sourceStat.mode
-    }));
+    prelayout.push(async () => {
```
Comment (seansfkelley) [inline .pnp.cjs:46004]: The build wanted me to `yarn install` but I don't understand why this change is here...

Comment (merceyz) [reply .pnp.cjs:46004]: The release workflow updated the file that this file is based on in https://github.com/yarnpkg/berry/commit/d7fecbed653462cf678ece387d8aba0f864ab6ab#diff-e2f3ae0b0dd7c014ad4e4fce4cb52742573bedb3761435d7f847570ab12730b9 but didn't commit it, the change to that function itself comes from https://github.com/yarnpkg/berry/pull/3338/files#diff-f9a18e543151537cdfdb81ff2877ba0d9c2c1c0876987e98c893b379455c2ec7

Comment (seansfkelley) [reply .pnp.cjs:46004]: Yeah, it looks like the master build is currently broken: https://github.com/yarnpkg/berry/runs/3595736129.

**Comment by arcanis** [comment]
Seems reasonable, thanks 👍

**Comment by seansfkelley** [comment]
Thanks for getting to it so quick!

**Comment by Psidium** [comment]
I came here looking for exactly this feature 👍

I'll open a PR updating the docs, if that's alright.
